Add a demo on Carstensen-Gedicke lower bounds for the Dirichlet Poisson eigenproblem (#5308)
firedrakeproject/firedrake: demos/poisson_adaptive_eigenvalue/poisson_adaptive_eigenvalue.py.rst

Adaptive eigenvalue problem on an L-shaped domain
=================================================

We consider the Dirichlet eigenvalue problem for the Poisson equation on an L-shaped domain :math:`\Omega`:

.. math::
   -\Delta u &= \lambda u \quad \text{in } \Omega, \\
           u &= 0         \quad \text{on } \partial \Omega.

> Since the domain has a re-entrant corner, the eigenfunctions exhibit a singularity. Thus, uniform mesh refinement leads to suboptimal convergence. We demonstrate an adaptive strategy driven by a residual-based a posteriori error estimator. To establish rigorous bounds on the true eigenvalue :math:`\lambda`, we use conforming elements (CG) for an upper bound, and nonconforming elements (CR) with Carstensen-Gedicke postprocessing :cite:`CarstensenGedicke:2014` for a guaranteed lower bound.

Reviewer: Would ne niece to also show the convergence rate of uniform refinement (should be h^{4/3}), to also graphically show that adaptivity outperforms uniform refinement at equal dof count.
What do you think ?

Author: That would be nice e.g. for a classroom example, but I think in these demos we should prioritise speed.

Author: On reflection I agree it is valuable, and it is only a small extra cost. I've added a comparison showing suboptimal convergence with uniform refinement.

> Since the domain has a re-entrant corner, the eigenfunctions exhibit a singularity. Thus, uniform mesh refinement leads to suboptimal convergence. We demonstrate an adaptive strategy driven by a residual-based a posteriori error estimator. To establish rigorous bounds on the true eigenvalue :math:`\lambda`, we use conforming elements (CG) for an upper bound, and nonconforming elements (CR) with Carstensen-Gedicke postprocessing :cite:`CarstensenGedicke:2014` for a guaranteed lower bound.

Reviewer: ```suggestion
Since the domain has a re-entrant corner, the eigenfunctions exhibit a singularity. Thus, uniform mesh refinement leads to suboptimal convergence. We demonstrate an adaptive strategy driven by a residual-based a posteriori error estimator. To establish rigorous bounds on the true eigenvalue :math:`\lambda`, we use conforming elements (CG) for an upper bound :cite:`Boffi:2010`, and nonconforming elements (CR) with Carstensen-Gedicke postprocessing :cite:`CarstensenGedicke:2014` for a guaranteed lower bound.
```

      dofs.append(uh.function_space().dim())
      print(f"Level {i}: Upper bound {lam_ub:.5f}, Lower bound {lam_lb:.5f}, Gap {err:.5e}")

      VTKFile(f"l_eigenfunction_{i}.pvd").write(uh)

      if err < 1e-2:
          break

      eta, _ = estimate_error(mesh, uh, lam_ub)
      mesh = adapt(mesh, eta)

We can plot the convergence of the error bound :math:`\lambda_{\text{ub}} - \lambda_{\text{lb}}` against the number of degrees of freedom. With adaptivity, we expect the optimal :math:`O(N^{-1})` convergence rate. ::

  try:
      import matplotlib.pyplot as plt
      import numpy as np

      plt.grid()
      plt.loglog(dofs, error_estimators, '-ok', label="Guaranteed error bound")
      scaling = 1.5 * error_estimators[0] / dofs[0]**-1
      plt.loglog(dofs, np.array(dofs)**(-1.0) * scaling, '--', label="Optimal convergence $N^{-1}$")
      plt.xlabel("Number of degrees of freedom $N$")
      plt.ylabel(r"Error bound $\lambda_{\text{ub}} - \lambda_{\text{lb}}$")
      plt.legend()
      plt.savefig("adaptive_eigenvalue_convergence.png")
  except ImportError:
      warning("Matplotlib not imported")

.. figure:: adaptive_eigenvalue_convergence.png
   :align: center

>    :figwidth: 80%

Reviewer 2: It'd be nice to show a picture of the mesh

Author: The mesh is very multiscale, with resolution focussed strongly at the re-entrant corner. A single image doesn't really capture it, and I don't think it's worth making a movie?

Reviewer 2: Maybe a sequence of pictures?

Author: Done!

  L = rect1 + rect2

  geo = OCCGeometry(L, dim=2)
  ngmesh = geo.GenerateMesh(maxh=0.5)
  mesh = Mesh(ngmesh)

We create a function to solve the eigenvalue problem for both continuous (CG) and nonconforming Crouzeix-Raviart (CR) :cite:`Crouzeix:1973` elements. By the Rayleigh-Ritz principle :cite:`Boffi:2010,Gander:2012`, the CG solution provides an upper bound :math:`\lambda_{\text{ub}}`. The CR solution gives a discrete eigenvalue :math:`\lambda_{\text{CR}}`, which can be postprocessed to yield a guaranteed lower bound:

.. math::
    \lambda_{\text{lb}} = \frac{\lambda_{\text{CR}}}{1 + \kappa_{\text{CR}}^2 h_{\max}^2 \lambda_{\text{CR}}}

where :math:`\kappa_{\text{CR}} \approx 0.1893` is the constant established by Carstensen and Gedicke :cite:`CarstensenGedicke:2014`. A general theory for deriving lower bounds for eigenvalues with nonconforming methods has been developed by Hu et al. :cite:`Hu:2014`. We will use the postprocessed lower bound to terminate the adaptive iteration, while for technical reasons we will plot the Galerkin gap :math:`\lambda_{\text{CG}} - \lambda_{\text{CR}}` to demonstrate optimal convergence. ::

>   def solve_poisson(mesh):

Reviewer 2: ```suggestion
  def solve_eigenproblem(mesh):
```

  def solve_eigenproblem(mesh):

>       h_max = Function(FunctionSpace(mesh, "DG", 0)).interpolate(CellDiameter(mesh)).dat.data_ro.max()

Reviewer 2: ```suggestion
      h_symbolic = CellDiameter(mesh)
      DG0 = FunctionSpace(mesh, "DG", 0)
      h_cell = Function(DG0).interpolate(h_symbolic)
      with h_cell.dat.vec_ro as hvec:
          h_max = hvec.max()[1]
```

      for space in ["CG", "CR"]:
          V = FunctionSpace(mesh, space, 1)

          u = TrialFunction(V)
          v = TestFunction(V)

          a = inner(grad(u), grad(v))*dx
          b = inner(u, v)*dx
          bc = DirichletBC(V, 0, "on_boundary")
          eigenproblem = LinearEigenproblem(a, b, bc)

          sp = {
                "eps_gen_hermitian": None,
                "eps_smallest_real": None,
                "eps_monitor_cancel": None,
                "eps_type": "krylovschur",
                "eps_target": 0,

>                 "st_type": "sinvert",

Reviewer 2: Does it fail with the default options? If it does these options should be explained

Author: It does fail with default options, so I've added an explanation.

Reviewer: Default options for our eigensolver, are not good in my very personal opinion.

                }

          eigensolver = LinearEigensolver(eigenproblem, 1, solver_parameters=sp)
          eigensolver.solve()

          if space == "CR":
              lambda_CR = eigensolver.eigenvalue(0).real
              kappa_CR = 0.1893
              bounds["lb"] = lambda_CR / (1 + kappa_CR**2 * h_max**2 * lambda_CR)
              bounds["CR"] = lambda_CR
          if space == "CG":
              bounds["ub"] = eigensolver.eigenvalue(0).real
              eigenfunction = eigensolver.eigenfunction(0)[0]

      eigenfunction.rename("Eigenfunction")

>       return (bounds["lb"], bounds["ub"], bounds["CR"], eigenfunction)

Reviewer 2: why do we need to keep a dict of bounds if we are going to unpack it anyway?

Author: Agreed, fixed

> These bounds do not describe where the mesh should be refined so as to reduce the error. For this purpose we employ a standard residual-based a posteriori error estimator :cite:`Duran:2003,Larson:2000`. Note that this assumes there is a single eigenfunction associated with the lowest eigenvalue; if the eigenvalue were of higher multiplicity the estimator would need to consider the entire eigenspace. ::

Reviewer: ```suggestion
These bounds do not describe where the mesh should be refined so as to reduce the error. For this purpose we employ a standard residual-based a posteriori error estimator :cite:`Duran:2003,Larson:2000`. Note that this assumes there is a single eigenfunction associated with the lowest eigenvalue; if the eigenvalue were of higher multiplicity the estimator would need to consider the entire eigenspace :cite:`Boffi:2014`. ::
```

  def estimate_error(mesh, uh, lam):
      W = FunctionSpace(mesh, "DG", 0)
      eta_sq = Function(W)
      w = TestFunction(W)
      h = CellDiameter(mesh)
      n = FacetNormal(mesh)
      v = CellVolume(mesh)

      G = (
            inner(eta_sq / v, w)*dx
          - inner(h**2 * (lam*uh + div(grad(uh)))**2, w) * dx
          - inner(h('+')/2 * jump(grad(uh), n)**2, w('+')) * dS
          - inner(h('-')/2 * jump(grad(uh), n)**2, w('-')) * dS
          )

>       sp = {"mat_type": "matfree", "ksp_type": "richardson", "pc_type": "jacobi"}

Reviewer 2: ```suggestion
      sp = {"mat_type": "matfree", "ksp_type": "preonly", "pc_type": "jacobi"}
```

      solve(G == 0, eta_sq, solver_parameters=sp)
      eta = Function(W).interpolate(sqrt(eta_sq))

      with eta.dat.vec_ro as eta_:

>           error_est = sqrt(eta_.dot(eta_))

Reviewer 2: ```suggestion
          error_est = eta_.norm()
```

      return (eta, error_est)

We define a function to adapt the mesh by refining elements with large error indicators, using the maximum Dörfler-like marking strategy with :math:`\theta = 0.5`: ::

  def adapt(mesh, eta):
      W = FunctionSpace(mesh, "DG", 0)
      markers = Function(W)

      with eta.dat.vec_ro as eta_:
          eta_max = eta_.max()[1]

      theta = 0.5
      should_refine = conditional(gt(eta, theta*eta_max), 1, 0)
      markers.interpolate(should_refine)

      return mesh.refine_marked_elements(markers)

Finally, we run the adaptive loop until the upper and lower bounds agree to within a tolerance. ::

  max_iterations = 100
  error_estimators = []
  dofs = []
  err = 1

  for i in range(max_iterations):

>       lam_lb, lam_ub, lam_CR, uh = solve_poisson(mesh)

Reviewer 2: ```suggestion
      lam_lb, lam_ub, lam_CR, uh = solve_eigenproblem(mesh)
```

>   max_iterations = 100

Reviewer 2: The demo only reaches level 15, we should not suggest that we are taking this many iterations.
```suggestion
  max_iterations = 20
```

      err = lam_ub - lam_lb
      gap = lam_ub - lam_CR
      error_estimators.append(gap)
      dofs.append(uh.function_space().dim())
      print(f"Level {i}: Upper bound {lam_ub:.5f}, Lower bound {lam_lb:.5f}, Bound gap {err:.5e}, Galerkin gap {gap:.5e}")

      VTKFile(f"l_eigenfunction_{i}.pvd").write(uh)

      if err < 1e-2 or dofs[-1] > 1000000:
          break

      eta, _ = estimate_error(mesh, uh, lam_ub)
      mesh = adapt(mesh, eta)

To demonstrate that adaptivity is necessary to achieve the optimal convergence rate, we can run the same script with :math:`\theta = 0`, which forces uniform refinement (all cells are marked for refinement at every step). We make this optional by guarding it behind the Boolean ``run_uniform``. ::

  run_uniform = True

  if run_uniform:
      mesh_uniform = Mesh(ngmesh)
      uniform_error_estimators = []
      uniform_dofs = []

      def adapt_uniform(mesh, eta):
          markers = Function(FunctionSpace(mesh, "DG", 0)).assign(1.0)
          return mesh.refine_marked_elements(markers)

      for i in range(max_iterations):

>           lam_lb, lam_ub, lam_CR, uh = solve_poisson(mesh_uniform)

Reviewer 2: ```suggestion
          lam_lb, lam_ub, lam_CR, uh = solve_eigenproblem(mesh_uniform)
```
